src/NServiceBus.Core.Tests/API/TaskReturning.cs

```diff
@@ -0,0 +1,287 @@
+﻿namespace NServiceBus.Core.Tests.API
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+
+    static class TaskReturning
+    {
+        static readonly List<MethodInfo> allMethods = typeof(IMessage).Assembly.GetTypes()
+            .Where(type => !type.IsObsolete())
+            .SelectMany(type => type.GetMethods(BindingFlags.Static | BindingFlags.Instance | BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.NonPublic))
+            .Where(method => !method.IsCompilerGenerated())
+            .Where(method => !method.IsObsolete())
+            .ToList();
+
+        static readonly List<MethodInfo> taskMethods = allMethods
+            .Where(method => typeof(Task).IsAssignableFrom(method.ReturnType))
+            .ToList();
+
+#pragma warning disable IDE0001 // Simplify Names
+        static readonly List<MethodInfo> noTokenPolicy = taskMethods
+            .Where(method =>
+                (typeof(Delegate).IsAssignableFrom(method.DeclaringType) && method.Name == "EndInvoke") ||
+                (method.HasCancellableContext() &&
+                    method.IsOnOneOf(
+                        typeof(NServiceBus.Saga),
+                        typeof(NServiceBus.IncomingMessageOperations),
+                        typeof(NServiceBus.IPipeline),
+                        typeof(NServiceBus.Pipeline.IBehavior),
+                        typeof(NServiceBus.IHandleMessages<>),
+                        typeof(NServiceBus.IHandleTimeouts<>),
+                        typeof(NServiceBus.Sagas.IHandleSagaNotFound),
+                        typeof(NServiceBus.Pipeline.MessageHandler),
+                        typeof(NServiceBus.IUnicastPublishRouter)
+                        )) ||
+                method.IsOn(typeof(NServiceBus.IAsyncTimer)) ||
```

Comment:
I'm having second thoughts about this one. I think, for safety, it may be better to add a mandatory token.

Reply:
I'm hoping this PR will make this a moot point.